src/btree/bt_vrfy.c

```diff
@@ -594,84 +588,98 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
-        WT_ERR(meta_cursor->get_value(meta_cursor, &meta_value));
-        if ((ret = __wt_config_getones(session, meta_value, "id", &id)) == 0 &&
-          btree_id == id.val) {
-            WT_ERR(meta_cursor->get_key(meta_cursor, uri));
-            break;
-        }
-        WT_ERR_NOTFOUND_OK(ret);
-    }
-err:
-    WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
-    return (ret);
-}
-
 /*
  * __wt_verify_history_store_tree --
- *     Verify the history store tree, by iterating and checking all the keys in the history and that
- *     it exists in both the meta file and the respective data store table. Skips duplicate keys by
- *     storing the previous key.
+ *     Verify the history store. There can't be an entry in the history store without having the
+ *     latest value for the respective key in the data store. If given a uri, limit the verification
+ *     to the corresponding btree.
  */
 int
-__wt_verify_history_store_tree(WT_SESSION_IMPL *session)
+__wt_verify_history_store_tree(WT_SESSION_IMPL *session, const char *uri)
 {
     WT_CURSOR *cursor, *data_cursor;
     WT_DECL_ITEM(tmp);
     WT_DECL_RET;
     WT_ITEM hs_key, prev_hs_key;
     WT_TIME_PAIR hs_start, hs_stop;
-    uint32_t btree_id, session_flags, prev_btree_id;
-    int cmp;
-    const char *uri;
+    uint32_t btree_id, btree_id_given_uri, session_flags, prev_btree_id;
+    int exact, cmp;
+    char *uri_itr;
 
     session_flags = 0;
-    prev_btree_id = 0; /* [-Wconditional-uninitialized] */
     data_cursor = NULL;
-    uri = NULL;
     WT_CLEAR(prev_hs_key);
     WT_CLEAR(hs_key);
+    btree_id_given_uri = 0; /* [-Wconditional-uninitialized] */
+    prev_btree_id = 0;      /* [-Wconditional-uninitialized] */
+    uri_itr = NULL;
 
     WT_ERR(__wt_scr_alloc(session, 0, &tmp));
     WT_ERR(__wt_hs_cursor(session, &session_flags));
     cursor = session->hs_cursor;
 
-    while ((ret = cursor->next(cursor)) == 0) {
+    /*
+     * If a uri has been provided, limit verification to the corresponding btree by jumping to the
+     * first record for that btree in the history store. Otherwise scan the whole history store.
+     */
+    if (uri != NULL) {
+        ret = __wt_metadata_uri_to_btree_id(session, uri, &btree_id_given_uri);
+        if (ret != 0)
+            WT_ERR_MSG(session, ret, "Unable to locate the URI %s in the metadata file.", uri);
+
+        /*
+         * Position the cursor at the first record of the specified btree, or one after. It is
+         * possible there are no records in the history store for this btree.
+         */
+        cursor->set_key(cursor, btree_id_given_uri, &hs_key, 0, 0, 0, 0);
+        ret = cursor->search_near(cursor, &exact);
+        if (ret == 0 && exact < 0)
+            ret = cursor->next(cursor);
+        if (ret == WT_NOTFOUND) {
+            ret = 0;
+            goto err;
+        }
+        WT_ERR(ret);
+    } else
+        ret = cursor->next(cursor);
```

Comment:
I think if you delete lines 638-642, it all still works, because `ret != 0` and you fall through the loop. Since that's what we're relying on for the `WT_CURSOR.next` call at line 644, might as well make both parts work the same?